[Complex text codepath] Use ICU to determine grapheme cluster boundaries

litherum opened the pull request:
#### 12bc2f0ab0d71a2a547031280df9eedf3c536a69
<pre>
[Complex text codepath] Use ICU to determine grapheme cluster boundaries
<a href="https://bugs.webkit.org/show_bug.cgi?id=243530">https://bugs.webkit.org/show_bug.cgi?id=243530</a>
&lt;rdar://problem/98103970&gt;

Reviewed by Yusuke Suzuki.

Instead of naive hand-rolled code that just checks for 4 specific classes of characters,
we should use ICU to correctly find the boundaries of grapheme clusters.

Because the complex text codepath is already so rare and slow, this patch has no effect on
performance (as measured using the Design MotionMark subtest).

Test: fast/text/thai-joiner.html

* Source/WTF/wtf/text/TextBreakIterator.cpp:
(WTF::mapModeToBackingIterator):
* Source/WTF/wtf/text/TextBreakIterator.h:
* Source/WTF/wtf/text/cocoa/TextBreakIteratorInternalICUCocoa.cpp:
(WTF::mapModeToBackingIterator):
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::advanceByCombiningCharacterSequence):
(WebCore::ComplexTextController::collectComplexTextRuns):

Canonical link: <a href="https://commits.webkit.org/253188@main">https://commits.webkit.org/253188@main</a>
</pre>

litherum commented on Source/WTF/wtf/text/TextBreakIterator.cpp:
```diff
@@ -52,6 +52,8 @@ static std::variant<TextBreakIteratorICU, TextBreakIteratorPlatform> mapModeToBa
     case TextBreakIterator::Mode::Delete:
         return TextBreakIteratorICU(string, TextBreakIteratorICU::Mode::Character, locale.string().utf8().data());
+    case TextBreakIterator::Mode::Character:
+        return TextBreakIteratorICU(string, TextBreakIteratorICU::Mode::Character, locale.string().utf8().data());
```
This should be using kCFStringComposedCharacterCluster instead

Constellation commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -371,9 +343,12 @@ void ComplexTextController::collectComplexTextRuns()
     const Font* synthesizedFont = nullptr;
     const Font* smallSynthesizedFont = nullptr;
 
+    CachedTextBreakIterator graphemeClusterIterator(m_run.text(), TextBreakIterator::Mode::Character, m_font.fontDescription().computedLocale());
```
While we are always using 16bit `UChar` iterator (by converting 8bit string to 16bit string), here we are passing StringView, which can be 8bit or 16bit. Don't we need to pass the same source to `curr` and `end`?

litherum replied:
`curr` and `end` are parallel to the `CachedTextBreakIterator`. They don't have to point to the same thing.

That being said, upon further inspection, I don't think we need `cp` or `curr` at all now. But that's a significant refactoring, so I'd like to do that in a follow-up patch.

Constellation reviewed (approved):
r=me with one comment.

litherum committed:
[Complex text codepath] Use ICU to determine grapheme cluster boundaries
https://bugs.webkit.org/show_bug.cgi?id=243530
<rdar://problem/98103970>

Reviewed by Yusuke Suzuki.

Instead of naive hand-rolled code that just checks for 4 specific classes of characters,
we should use ICU to correctly find the boundaries of grapheme clusters.

Because the complex text codepath is already so rare and slow, this patch has no effect on
performance (as measured using the Design MotionMark subtest).

Test: fast/text/thai-joiner.html

* Source/WTF/wtf/text/TextBreakIterator.cpp:
(WTF::mapModeToBackingIterator):
* Source/WTF/wtf/text/TextBreakIterator.h:
* Source/WTF/wtf/text/cocoa/TextBreakIteratorInternalICUCocoa.cpp:
(WTF::mapModeToBackingIterator):
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::advanceByCombiningCharacterSequence):
(WebCore::ComplexTextController::collectComplexTextRuns):

Canonical link: https://commits.webkit.org/253188@main
